index.rst

```diff
@@ -55,78 +18,545 @@
+
+.. note::
+
+    Once the ``Butler`` redesign is complete and the :py:class:`SuperTask` design has been updated accordingly, this document will be deprecated and its content moved to `LDM-152 <https://ldm-152.lsst.io>`_, the LSST DM Middleware Design Document.
+
+
 .. _task_config_context:
 
 Task/Config Context
 ===================
 
+The design for the SuperTask Library sits on top of the existing LSST concepts of *Task* and *Config* classes.
+All SuperTasks are Tasks, and utilize the same Config system for algorithmic configuration parameters.
+
+A concrete Task is simply a configurable, composable, callable object.
+While all Tasks inherit from a common abstract base class and define a ``run`` method, each Task defines its own signature for ``run``, so Tasks do not really share a common interface.
+The ``lsst.pipe.base.Task`` abstract base class itself exists largely to provide utility code to its subclasses, such as setting up logging, providing objects to hold processing metadata, and setting up nested Tasks to be delegated to, called *subtasks*.
+
+This composition of Tasks is closely tied to our approach for configuring them, and it is this functionality that makes the Task concept so useful.
+Configuration options for a Task are defined by a corresponding class that inherits from ``lsst.pex.config.Config``.
+Config classes contain ``lsst.pex.config.Field`` instances, which act like introspectable properties that define the types, allowed values, and documentation for configuration options.
+A set of configuration values for a Task is thus just an instance of its Config class, and overrides for those values can be expressed as Python code that assigns values to to the ``Field`` attributes.
+When a Task delegates to another as a subtask, its Config class usually contains a special ``ConfigurableField`` that holds an instance of the subtask's Config class.
+``ConfigurableField`` allows this Config instance to be replaced by one for a different Task, allowing the subtask to be replaced by another with the same ``run`` signature.
+
+The :py:class:`SuperTask` abstract base class inherits from ``Task``, and its concrete subclasses are expected to defined a Config class to define their configuration parameters and delegate additional work to subtasks.
+Using a SuperTask *as* a subtask is not meaningful, however; in that context the SuperTask just behaves like a regular Task and the additional interfaces and functionality added by SuperTask go unused (as a result, we expect this to be rare).
+
+A few additional properties of Tasks are particularly relevant for SuperTask design:
+
+- The configuration of a Task is frozen after the Task is constructed.
+
+- The schema of any catalogs produced by a Task must be fully defined after Task construction, and must not depend on the actual contents of any data products.
+
+- Calls to ``run`` or any other methods must not change any internal state.
```

Comment:
Sorry for coming in so late, but what does this requirement imply for, e.g., task metadata? Currently metadata is implemented using a mutable member of ``Task``, which violates this rule.

Reply:
Good catch; Task metadata has always been a weird outlier in this context, and I *think* there's some magic in `CmdLineTask` somewhere to make sure we actually always get a fresh metadata object on each call to the `CmdLineTask`'s `run`.  I'd like to find some other way to make metadata output available to Tasks to remove that special treatment, but haven't thought deeply about how to do it.

Reply:
As far as I can tell there's no magic for this in `CmdLineTask`. The default `TaskRunner` sidesteps the problem by creating a new Task object for each unit of work; presumably runners that call `run` multiple times on the same object need to deal with the consequences themselves.

Reply:
Ah, good to know.  I'm still fairly uncomfortable with that arrangement, but it is a bit safer than what I thought was going on.